onnxruntime/core/framework/graph_partitioner.cc

```diff
@@ -848,20 +854,39 @@ static Status CreateEpContextModel(const ExecutionProviders& execution_providers
     }
   }
 
-  size_t ini_size_threshold = 0;
-  std::filesystem::path external_ini_path;
-  if (ep_context_ext_ini_path.empty()) {
+  size_t ini_size_threshold = ep_context_gen_options.output_external_initializer_size_threshold;
+  std::filesystem::path external_ini_path = ep_context_gen_options.output_external_initializers_file_path;
+  if (external_ini_path.empty()) {
     // Set the threshold to the max so all initializers are forced into the Onnx file
     ini_size_threshold = SIZE_MAX;
     external_ini_path = "./model_ext_ini.bin";
-  } else {
-    // Set the theshold to 0 so all initializers are forced into the external file
-    ini_size_threshold = 0;
-    external_ini_path = ep_context_ext_ini_path;
   }
+
   ModelSavingOptions model_saving_options{ini_size_threshold};
-  ORT_RETURN_IF_ERROR(Model::SaveWithExternalInitializers(ep_context_model, context_cache_path,
-                                                          external_ini_path, model_saving_options));
+
+  if (ep_context_gen_options.output_model_buffer_ptr != nullptr &&
+      ep_context_gen_options.output_model_buffer_size_ptr != nullptr &&
+      ep_context_gen_options.output_model_buffer_allocator != nullptr) {
+    ORT_RETURN_IF_ERROR(ep_context_model.MainGraph().Resolve());
+    // TODO(adrianlizarraga): Investigate if we can make this more memory efficient.
+    // May be able to use allocator to directly allocate the ModelProto to avoid a copy.
+    ONNX_NAMESPACE::ModelProto model_proto = ep_context_model.ToGraphProtoWithExternalInitializers(external_ini_path,
+                                                                                                   context_cache_path,
+                                                                                                   model_saving_options);
+    size_t buffer_size = model_proto.ByteSizeLong();
+    ORT_RETURN_IF(buffer_size > static_cast<size_t>(std::numeric_limits<int>::max()),
+                  "Cannot serialize ONNX ModelProto larger than 2GB");
+
+    OrtAllocator* allocator = ep_context_gen_options.output_model_buffer_allocator;
```

Comment:
>OrtAllocator* allocator [](http://example.com/codeflow?start=4&length=23)

Typically, we wrap the incoming OrtAllocator pointers into a special Wrapper class, so it is represented as AllocatorPtr. This can be used in other places in our code safely and it can be used within a smart ptr.

Perhaps, we need to check if storing a raw ptr here is a good idea.